qip-0014.md

```diff
@@ -0,0 +1,170 @@
+```
+ QIP: 14
+ Layer: EVM
+ Title: CREATE + CREATE2 opcode
+ Author: alanorwick <alan+qip@quai.org>
+ Comments-Summary: No comments yet.
+ Comments-URI: https://github.com/quainetwork/qips/wiki/Comments:QIP-0014
+ Status: Draft
+ Type: Informational
+ Created: 2024-07-31
+ License: BSD-2-Clause
+```
+
+
+## Abstract
+
+This QIP proposes modifications to the CREATE opcode in Quai Network to ensure shard-specific address generation, while maintaining CREATE2 functionality as is. The proposal aims to adapt Ethereum's address creation mechanism to Quai's sharded architecture.
+
+## Motivation
+
+Quai Network's sharded architecture requires addresses to be shard-specific. The current Ethereum-style CREATE opcode does not guarantee this, potentially leading to cross-shard conflicts and inefficiencies. This proposal seeks to address this issue while maintaining as much compatibility with Ethereum as possible.
+
+## Specification
+
+### CREATE Opcode Modification
+
+1. The CREATE opcode will generate an address using the following method:
+   
+   a. Split the bytecode_data into two parts:
+      - `hashCode`: All but the last 4 bytes of bytecode_data
+      - `potentialSalt`: The last 4 bytes of bytecode_data
+   
+   b. Calculate the initial address:
+
+      ```address = keccak256(([sender_address, sender_nonce, keccak256(hashCode), potentialSalt])[12:]```
+   
+   c. If the resulting address is valid for the current shard:
+      - The full bytecode_data (including the last 4 bytes) is used as the contract bytecode
+      - The `potentialSalt` is considered part of the actual bytecode
+   
+   d. If the resulting address is not valid for the current shard:
+      - Proceed to step 2 for grinding
+
+2. If the resulting address is not valid for the current shard:
+The gas cost per iteration of bytecode modification and checking is the `KeccackGasCost` of the size of `concat(sender_address, sender_nonce, hashCodeHash, salt)`.
+  
+   a. Initialize a "grind nonce" to 0.
```

Comment:
Why init to zero? Should start with the nonce the user already provided.

Reply:
The salt is assumed to be part of the bytecode unless the user provided one is not valid. This QIP defines the first check of the salt potentially be part of the bytecode itself. The salt is then appended if the prior four bytes (bytecode or extra data) does not result in a valid address.

Reply:
> The salt is assumed to be part of the bytecode

This is not a good idea. When I interpret contract data later, how am I supposed to know if the last 4 bytes are actual contract data, or if they are just salt for the address? This could potentially lead to issues, if a contract was interpreted one way at deployment, but then interpreted differently when it is executed.

The salt needs to be explicitly specified for every contract.